pkg/server/server.go

```diff
@@ -517,3 +540,63 @@ func goContext(parent genericapiserver.PostStartHookContext) context.Context {
 func (s *Server) WaitForPhase1Finished() {
 	<-s.rootPhase1FinishedCh
 }
+
+func (s *Server) leaderElectAndRun(ctx context.Context) {
+	logger := klog.FromContext(ctx)
+
+	id, err := os.Hostname()
+	if err != nil {
+		logger.Error(err, "failed to determine hostname")
+		return
+	}
+
+	id = id + "_" + string(uuid.NewUUID())
+
+	rl, err := resourcelock.NewFromKubeconfig("leases", metav1.NamespaceSystem, "kcp-controllers",
+		resourcelock.ResourceLockConfig{
+			Identity: id,
+		},
+		s.ExtraConfig.LeaderElectionConfig,
+		time.Second*30,
+	)
+
+	if err != nil {
+		logger.Error(err, "failed to take resourcelock")
+		return
+	}
+
+	logger = logger.WithValues("identity", id)
+	logger.Info("starting leader election")
+
+	leaderelection.RunOrDie(ctx, leaderelection.LeaderElectionConfig{
+		Lock:          rl,
+		LeaseDuration: time.Second * 60,
+		RenewDeadline: time.Second * 5,
+		RetryPeriod:   time.Second * 2,
+		Callbacks: leaderelection.LeaderCallbacks{
+			OnStartedLeading: func(ctx context.Context) {
+				logger.Info("started leading")
+
+				for controllerName, controller := range s.controllers {
+					logger := logger.WithValues("controller", controllerName)
+					go s.runController(ctx, controller, logger)
+				}
+			},
+			OnStoppedLeading: func() {
+				logger.Info("stopped leading")
+			},
```

Comment:
I'm trying to work through the scenario which looks like:
```
starts leading -> stops leading -> starts leading
```
If this were to happen, we end up re-spawning all registered controllers.

Reply:
My understanding - and it might be flawed - is that all controllers should terminate because the context passed to them is canceled upon losing a leader election. So the idea for me was that re-spawning all controllers is exactly what we want when winning a leader election.

However, I haven't gotten around to actually test this yet. So I could be wrong here.